gdk4/src/subclass/content_provider.rs

```diff
@@ -0,0 +1,433 @@
+                .to_glib_none()
+                .0);
+
+            from_glib_full(ret)
+        }
+    }
+
+    fn parent_write_mime_type_async<
+            P: IsA<gio::OutputStream>,
+            Q: IsA<gio::Cancellable>,
+            R: FnOnce(Result<(), glib::Error>) + Send + 'static,
+    >(
+        &self,
+        provider: &Self::Type,
+        mime_type: &str,
+        stream: &P,
+        io_priority: glib::Priority,
+        cancellable: Option<&Q>,
+        callback: R,
+    ) {
+        unsafe {
+            let data = T::type_data();
+            let parent_class = data.as_ref().parent_class() as *mut ffi::GdkContentProviderClass;
+            let f = (*parent_class)
+                .write_mime_type_async
+                .expect("no parent \"write_mime_type_async\" implementation");
+            let finish = (*parent_class)
+                .write_mime_type_finish
+                .expect("no parent \"write_mime_type_finish\" implementation");
+
+            let user_data: Box<(R, _)> = Box::new((callback, finish));
+
+            extern "C" fn parent_write_mime_type_async_trampoline<
+                R: FnOnce(Result<(), glib::Error>) + Send + 'static
+            >(
+                source_object_ptr: *mut glib::gobject_ffi::GObject,
+                res: *mut gio::ffi::GAsyncResult,
+                user_data: glib::ffi::gpointer,
+            ) {
+                let mut error = std::ptr::null_mut();
+                let cb: Box<(R, fn(*mut ffi::GdkContentProvider, *mut gio::ffi::GAsyncResult, *mut *mut glib::ffi::GError) -> i32)> = Box::from_raw(user_data as *mut _);
+                let ptr = cb.1(
+                    source_object_ptr as _,
+                    res,
+                    &mut error,
+                );
+                let result = if error.is_null() {
+                    Ok(())
+                } else {
+                    Err(from_glib_full(error))
+                };
+                cb.0(result);
+            }
+
+            let cancellable = cancellable.map(|p| p.as_ref());
+            let callback = parent_write_mime_type_async_trampoline::<R>;
+            f(
+                provider
+                    .unsafe_cast_ref::<ContentProvider>()
+                    .to_glib_none()
+                    .0,
+                mime_type.to_glib_none().0,
+                stream.as_ref().to_glib_none().0,
+                io_priority.to_glib(),
+                cancellable.to_glib_none().0,
+                Some(callback),
+                Box::into_raw(user_data) as *mut _,
+            );
+        }
+    }
+
+    fn parent_write_mime_type_future<P: IsA<gio::OutputStream> + Clone + 'static>(
+        &self,
+        provider: &Self::Type,
+        mime_type: &str,
+        stream: &P,
+        io_priority: glib::Priority,
+    ) -> Pin<Box<dyn Future<Output = Result<(), glib::Error>> + 'static>> {
+        let stream = stream.clone();
+        Box::pin(gio::GioFuture::new(provider, move |obj, send| {
```

Comment:
You have to clone the provider and everything. You can only have `'static` types inside that closure.

However the `_future` variant of the async functions should only be in the non-subclass API. It's a wrapper around the `_async` functions for user convenience.

Reply:
Yeha, it compiles. No idea if it works though :)

Right about parent_write_mime_type_future, but I thought ContentProviderImplExt was the right place. Where else?

Reply:
Ah so you want the user to implement the futures version and not the callback-based one. I guess that works.

Reply:
Well, implement it somewhere and make sure that it actually works :)

Reply:
Would be nice to have it as an example as well if possible

Reply:
seems to do the job